data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -541,6 +548,105 @@ private StateSyncCommitteesData getSyncCommitteesFromState(
             committeeIndices, spec.atEpoch(epoch).getConfig().getTargetCommitteeSize()));
   }
 
+  public SafeFuture<Optional<SyncCommitteeRewardData>> getSyncCommitteeRewardsFromBlockId(
+      final String blockId, final Set<String> validators) {
+    return getBlockAndMetaData(blockId)
+        .thenCompose(
+            result -> {
+              if (result.isEmpty() || result.get().getData().getBeaconBlock().isEmpty()) {
+                return SafeFuture.completedFuture(Optional.empty());
+              }
+
+              final BlockAndMetaData blockAndMetaData = result.get();
+              final BeaconBlock block = blockAndMetaData.getData().getBeaconBlock().get();
+              final SyncCommitteeRewardData data =
+                  new SyncCommitteeRewardData(
+                      blockAndMetaData.isExecutionOptimistic(), blockAndMetaData.isFinalized());
+
+              return combinedChainDataClient
+                  .getStateByBlockRoot(block.getRoot())
+                  .thenApply(
+                      maybeState -> {
+                        if (maybeState.isEmpty()) {
+                          return Optional.empty();
+                        }
+
+                        final tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState
```

Comment:
Potentially a third private function in here, which is the business type logic...
You could then test scenarios like missing keys or indices out of bounds and push what happens in unit tests very easily given a state and block